osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using NUnit.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Testing;
+using osu.Game.Beatmaps;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Rulesets;
+using osu.Game.Screens.Select;
+using osu.Game.Tests.Visual.Navigation;
+using osu.Game.Users;
+
+namespace osu.Game.Tests.Visual.SongSelect
+{
+    [HeadlessTest]
+    public class TestSceneBeatmapRecommendations : OsuGameTestScene
+    {
+        [Resolved]
+        private DifficultyRecommender recommender { get; set; }
+
+        [Resolved]
+        private RulesetStore rulesets { get; set; }
+
+        [SetUpSteps]
+        public new void SetUpSteps()
+        {
+            AddStep("register request handling", () =>
+            {
+                ((DummyAPIAccess)API).HandleRequest = req =>
+                {
+                    switch (req)
+                    {
+                        case GetUserRequest userRequest:
+
+                            decimal pp = userRequest.Ruleset.ID switch
+                            {
+                                0 => 336, // Expected recommended star difficulty 2*
+                                1 => 928, // Expected recommended star difficulty 3*
+                                2 => 1905, // Expected recommended star difficulty 4*
+                                3 => 3329, // Expected recommended star difficulty 5*
+                                _ => 0
+                            };
+
+                            userRequest.TriggerSuccess(new User
+                            {
+                                Username = @"Dummy",
+                                Id = 1001,
+                                Statistics = new UserStatistics
+                                {
+                                    PP = pp
+                                }
+                            });
+                            break;
+                    }
+                };
+                // Force recommender to calculate its star ratings again
+                recommender.APIStateChanged(API, APIState.Online);
+            });
+        }
+
+        [Test]
+        public void TestPresentedBeatmapIsRecommended()
+        {
+            var importFunctions = new List<Func<BeatmapSetInfo>>();
+
+            for (int i = 0; i < 5; i++)
+            {
+                importFunctions.Add(importBeatmap(i, new List<RulesetInfo> { null, null, null, null, null }));
+            }
+
+            for (int i = 0; i < 5; i++)
+            {
+                presentAndConfirm(importFunctions[i], i, 2);
+            }
+        }
+
+        [Test]
+        public void TestBestRulesetIsRecommended()
+        {
+            var osuRuleset = rulesets.AvailableRulesets.First(r => r.ID == 0);
```

Comment:
Generally these should be `Single`s, but you can actually simply do
```suggestion
            var osuRuleset = rulesets.GetRuleset(0);
```
(not going to point out every single instance).